# **<font color='white gray'>panData</font>**
# **<font color='white gray'>Data Analysis Project with Python Language</font>**
### <font color='white gray'>Data Cleaning and Missing Value Treatment Techniques for Data Analysis</font>
### **<font color='blue'>Part 1</font>**

## **Python Packages Used in the Project**

In [40]:
!pip install -q -U watermark

In [41]:
# Imports
import math
import sys, os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import warnings
warnings.filterwarnings('ignore')

In [42]:
%reload_ext watermark
%watermark -a "panData"

Author: panData



## **Loading the Data**

https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

In [43]:
# 1. We create a list to identify missing values
missing_values_labels_list = ["n/a", "na", "undefined"]

In [44]:
# 2. Load the dataset
dataset = pd.read_csv("dataset.csv", na_values=missing_values_labels_list)

In [45]:
# 3. Shape
dataset.shape

(150001, 55)

In [46]:
# 4. Data Sample
dataset.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  9.16456699548519E+015             42.0              5.0   
1                L77566A             65.0              5.0   
2                D42335A              NaN              NaN   
3                T21824A              NaN              NaN   
4                D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0              Samsung   Samsung Galaxy A5 Sm-A520F   
1   

In [47]:
# 5. Set the total number of columns to display when printing the dataframe
pd.set_option('display.max_columns', 100)

In [48]:
# 6. Data Sample
dataset.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  9.16456699548519E+015             42.0              5.0   
1                L77566A             65.0              5.0   
2                D42335A              NaN              NaN   
3                T21824A              NaN              NaN   
4                D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0              Samsung   Samsung Galaxy A5 Sm-A520F   
1   

In [49]:
# 7. Loading the data dictionary
dictionary = pd.read_excel("dictionary.xlsx")

In [50]:
# 8. Shape
dictionary.shape

(56, 2)

In [51]:
# 9. Data Sample
dictionary.head(10)

Fields  \
0      bearer id   
1      Dur. (ms)   
2          Start   
3       Start ms   
4            End   
5         End ms   
6       Dur. (s)   
7           IMSI   
8  MSISDN/Number   
9           IMEI   

                                                             Description  
0                                                 xDr session identifier  
1                                      Total Duration of the xDR (in ms)  
2                          Start time of the xDR (first frame timestamp)  
3  Milliseconds offset of start time for the xDR (first frame timestamp)  
4                             End time of the xDR (last frame timestamp)  
5      Milliseconds offset of end time of the xDR (last frame timestamp)  
6                                       Total Duration of the xDR (in s)  
7                               International Mobile Subscriber Identity  
8          MS International PSTN/ISDN Number of mobile - customer number  
9                                International Mobile Equipment Identity

In [52]:
# 10. Set a large value for the maximum column width
pd.set_option('display.max_colwidth', 100)

In [53]:
# 11. Data Sample
dictionary.head(60)

Fields  \
0                                  bearer id   
1                                  Dur. (ms)   
2                                      Start   
3                                   Start ms   
4                                        End   
5                                     End ms   
6                                   Dur. (s)   
7                                       IMSI   
8                              MSISDN/Number   
9                                       IMEI   
10                        Last Location Name   
11                           Avg RTT DL (ms)   
12                           Avg RTT UL (ms)   
13                   Avg Bearer TP DL (kbps)   
14                   Avg Bearer TP UL (kbps)   
15               TCP DL Retrans. Vol (Bytes)   
16               TCP UL Retrans. Vol (Bytes)   
17                       DL TP < 50 Kbps (%)   
18            50 Kbps < DL TP < 250 Kbps (%)   
19             250 Kbps < DL TP < 1 Mbps (%)   
20                        DL TP > 1 Mbps (%)   
21                       UL TP < 10 Kbps (%)   
22             10 Kbps < UL TP < 50 Kbps (%)   
23            50 Kbps < UL TP < 300 Kbps (%)   
24                      UL TP > 300 Kbps (%)   
25                           HTTP DL (Bytes)   
26                           HTTP UL (Bytes)   
27                 Activity Duration DL (ms)   
28                 Activity Duration UL (ms)   
29                               Dur. (ms).1   
30                      Handset Manufacturer   
31                              Handset Type   
32           Nb of sec with 125000B < Vol DL   
33     Nb of sec with 1250B < Vol UL < 6250B   
34  Nb of sec with 31250B < Vol DL < 125000B   
35            Nb of sec with 37500B < Vol UL   
36    Nb of sec with 6250B < Vol DL < 31250B   
37    Nb of sec with 6250B < Vol UL < 37500B   
38             Nb of sec with Vol DL < 6250B   
39             Nb of sec with Vol UL < 1250B   
40                   Social Media DL (Bytes)   
41                   Social Media UL (Bytes)   
42                        YouTube DL (Bytes)   
43                        YouTube UL (Bytes)   
44                        Netflix DL (Bytes)   
45                        Netflix UL (Bytes)   
46                         Google DL (Bytes)   
47                         Google UL (Bytes)   
48                          Email DL (Bytes)   
49                          Email UL (Bytes)   
50                         Gaming DL (Bytes)   
51                         Gaming UL (Bytes)   
52                                  Other DL   
53                                  Other UL   
54                          Total DL (Bytes)   
55                          Total UL (Bytes)   

                                                                            Description  
0                                                                xDr session identifier  
1                                                     Total Duration of the xDR (in ms)  
2                                         Start time of the xDR (first frame timestamp)  
3                 Milliseconds offset of start time for the xDR (first frame timestamp)  
4                                            End time of the xDR (last frame timestamp)  
5                     Milliseconds offset of end time of the xDR (last frame timestamp)  
6                                                      Total Duration of the xDR (in s)  
7                                              International Mobile Subscriber Identity  
8                         MS International PSTN/ISDN Number of mobile - customer number  
9                                               International Mobile Equipment Identity  
10                          User location call name (2G/3G/4G) at the end of the bearer  
11                     Average Round Trip Time measurement Downlink direction (msecond)  
12                       Average Round Trip Time measurement Uplink direction (msecond)  
13                Average Bearer Throughp

## **Exploratory Analysis**

In [54]:
# 12. Info
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

In [55]:
# 13. Descriptive Statistics
dataset.describe()

Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.490100e+05  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082016e+14   
std    2.893173e+18     288.611834     288.097653  8.103762e+04  2.148809e+10   
min    6.917538e+18       0.000000       0.000000  7.142000e+03  2.040471e+14   
25%    7.349883e+18     250.000000     251.000000  5.744050e+04  2.082014e+14   
50%    7.349883e+18     499.000000     500.000000  8.639900e+04  2.082015e+14   
75%    1.304243e+19     749.000000     750.000000  1.324302e+05  2.082018e+14   
max    1.318654e+19     999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847455e+13       109.795706        17.662883   
std     2.447443e+12  2.241637e+13       619.782739        84.793524   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365130e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611970e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
count            150000.000000            150000.000000   
mean              13300.045927              1770.428647   
std               23971.878541              4625.355500   
min                   0.000000                 0.000000   
25%                  43.000000                47.000000   
50%                  63.000000                63.000000   
75%               19710.750000              1120.000000   
max              378160.000000             58613.000000   

       TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
count                 6.185500e+04                 5.335200e+04   
mean                  2.080991e+07                 7.596587e+05   
std                   1.825665e+08                 2.645305e+07   
min                   2.000000e+00                 1.000000e+00   
25%                   3.565150e+04                 4.694750e+03   
50%                   5.687300e+05                 2.094950e+04   
75%                   3.768308e+06                 8.402025e+04   
max                   4.294426e+09                 2.908226e+09   

       DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
count        149247.000000                   149247.000000   
mean             92.844754                        3.069355   
std              13.038031                        6.215233   
min               0.000000                        0.000000   
25%              91.000000                        0.000000   
50%             100.000000                        0.000000   
75%             100.000000                        4.000000   
max             100.000000                       93.000000   

       250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
count                  149247.000000       149247.000000        149209.000000   
mean                        1.717341            1.609654            98.530142   
std                         4.159538            4.828890             4.634285   
min                         0.000000            0.000000             0.000000   
25%                         0.000000            0.000000            99.000000   
50%                         0.000000            0.000000           100.000000   
75%                         1.000000            0.000000           100.000000   
max                       100.000000           94.000000           100.000000   

       10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
count                  149209.000000                   149209.000000   
mean                        0.776749                        0.

It doesn’t make sense to calculate descriptive statistics for Beared Id, IMSI, MSISDN / Number, and IMEI. However, the describe() method calculates statistics for all numeric columns. These statistics are being computed before the data is cleaned, so there may be changes after missing values and outliers are addressed.

In [56]:
# 14. Shape
dataset.shape

(150001, 55)

In [57]:
# 15. Shape
dictionary.shape

(56, 2)

There are 150,001 rows and 55 columns in the dataframe. However, we have 56 columns with their names and descriptions in the dictionary. This means there is a described column that is not included in the dataframe. Let’s identify the missing column.

In [58]:
# 16. Concatenate the dataframes
df_column_comparison = pd.concat([pd.Series(dataset.columns.tolist()), dictionary['Fields']],
                                 axis=1)

In [59]:
# 17. Column names
df_column_comparison.columns

Index([0, 'Fields'], dtype='object')

In [60]:
# 18. Rename columns
df_column_comparison.rename(columns={0: 'Column in Dataset', 'Fields': 'Column in Dictionary'},
                            inplace=True)

In [61]:
# 19. View
df_column_comparison

Column in Dataset  \
0                                  Bearer Id   
1                                      Start   
2                                   Start ms   
3                                        End   
4                                     End ms   
5                                  Dur. (ms)   
6                                       IMSI   
7                              MSISDN/Number   
8                                       IMEI   
9                         Last Location Name   
10                           Avg RTT DL (ms)   
11                           Avg RTT UL (ms)   
12                   Avg Bearer TP DL (kbps)   
13                   Avg Bearer TP UL (kbps)   
14               TCP DL Retrans. Vol (Bytes)   
15               TCP UL Retrans. Vol (Bytes)   
16                       DL TP < 50 Kbps (%)   
17            50 Kbps < DL TP < 250 Kbps (%)   
18             250 Kbps < DL TP < 1 Mbps (%)   
19                        DL TP > 1 Mbps (%)   
20                       UL TP < 10 Kbps (%)   
21             10 Kbps < UL TP < 50 Kbps (%)   
22            50 Kbps < UL TP < 300 Kbps (%)   
23                      UL TP > 300 Kbps (%)   
24                           HTTP DL (Bytes)   
25                           HTTP UL (Bytes)   
26                 Activity Duration DL (ms)   
27                 Activity Duration UL (ms)   
28                               Dur. (ms).1   
29                      Handset Manufacturer   
30                              Handset Type   
31           Nb of sec with 125000B < Vol DL   
32     Nb of sec with 1250B < Vol UL < 6250B   
33  Nb of sec with 31250B < Vol DL < 125000B   
34            Nb of sec with 37500B < Vol UL   
35    Nb of sec with 6250B < Vol DL < 31250B   
36    Nb of sec with 6250B < Vol UL < 37500B   
37             Nb of sec with Vol DL < 6250B   
38             Nb of sec with Vol UL < 1250B   
39                   Social Media DL (Bytes)   
40                   Social Media UL (Bytes)   
41                         Google DL (Bytes)   
42                         Google UL (Bytes)   
43                          Email DL (Bytes)   
44                          Email UL (Bytes)   
45                        Youtube DL (Bytes)   
46                        Youtube UL (Bytes)   
47                        Netflix DL (Bytes)   
48                        Netflix UL (Bytes)   
49                         Gaming DL (Bytes)   
50                         Gaming UL (Bytes)   
51                          Other DL (Bytes)   
52                          Other UL (Bytes)   
53                          Total UL (Bytes)   
54                          Total DL (Bytes)   
55                                       NaN   

                        Column in Dictionary  
0                                  bearer id  
1                                  Dur. (ms)  
2                                      Start  
3                                   Start ms  
4                                        End  
5                                     End ms  
6                                   Dur. (s)  
7                                       IMSI  
8                              MSISDN/Number  
9                                       IMEI  
10                        Last Location Name  
11                           Avg RTT DL (ms)  
12                           Avg RTT UL (ms)  
13                   Avg Bearer TP DL (kbps)  
14                   Avg Bearer TP UL (kbps)  
15               TCP DL Retrans. Vol (Bytes)  
16               TCP UL Retrans. Vol (Bytes)  
17                       DL TP < 50 Kbps (%)  
18            50 Kbps < DL TP < 250 Kbps (%)  
19             250 Kbps < DL TP < 1 Mbps (%)  
20                        DL TP > 1 Mbps (%)  
21                       UL TP < 10 Kbps (%)  
22             10 Kbps < UL TP < 50 Kbps (%)  
23            50 Kbps < UL TP < 300 Kbps (%)  
24                      UL TP > 300 Kbps (%)  
25                           HTTP DL (Bytes)  
26                   

“Dur. (Ms)” is missing in the dataset, as seen at index 1 in df_column_comparison. This is where the order of columns started to shift.

However, the same column name, “Dur. (Ms),” appears in the dataset at index 5, while the dictionary file labels it as “Dur. (S)” at index 6. Since the units differ between these columns, as indicated in their names, we need to verify which one is correct. To investigate further, we will use the “Dur. (Ms) .1” column, located at indices 28 and 29 in the dataset and dictionary file, respectively.

In [62]:
# 20. Select columns for investigation
dataset[['Dur. (ms)', 'Dur. (ms).1']]

Dur. (ms)   Dur. (ms).1
0       1823652.0  1.823653e+09
1       1365104.0  1.365104e+09
2       1361762.0  1.361763e+09
3       1321509.0  1.321510e+09
4       1089009.0  1.089009e+09
...           ...           ...
149996    81230.0  8.123076e+07
149997    97970.0  9.797070e+07
149998    98249.0  9.824953e+07
149999    97910.0  9.791063e+07
150000        NaN           NaN

[150001 rows x 2 columns]

It appears that the “Dur. (Ms)” column is measured in seconds. Therefore, let’s rename it accordingly. We’ll also rename some other columns to make them clearer according to their descriptions and to align with the naming style of other columns.

In [63]:
# 21. Rename columns
dataset.rename(columns={'Dur. (ms)': 'Dur (s)',
                        'Dur. (ms).1': 'Dur (ms)',
                        'Start ms': 'Start Offset (ms)',
                        'End ms': 'End Offset (ms)'},
               inplace=True)

In [64]:
# 22. List of dataset columns
dataset.columns.tolist()

['Bearer Id',
 'Start',
 'Start Offset (ms)',
 'End',
 'End Offset (ms)',
 'Dur (s)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur (ms)',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 12

In [65]:
# 23. Shape
dataset.shape

(150001, 55)

## **Step 1 - Handling Missing Values**
1. Identifying Missing Values
2. Dropping Columns
3. Imputation with Backward Fill
4. Imputation with Forward Fill
5. Imputation of Categorical Variables
6. Dropping Rows

### **1.1. Identifying Missing Values**

In [66]:
# 24. Function to calculate the percentage of missing values
def calculate_missing_values_percentage(df):

    # Calculate the total number of cells in the dataset
    totalCells = np.product(df.shape)

    # Count the number of missing values per column
    missingCount = df.isnull().sum()

    # Calculate the total number of missing values
    totalMissing = missingCount.sum()

    # Calculate the percentage of missing values
    print("The dataset has", round(((totalMissing / totalCells) * 100), 2), "%", "missing values.")

In [67]:
# 25. Check the percentage of missing values
calculate_missing_values_percentage(dataset)

The dataset has 12.72 % missing values.


In [68]:
# 26. Function to calculate missing values by column
def calculate_missing_values_by_column(df):

    # Total missing values
    mis_val = df.isnull().sum()

    # Percentage of missing values
    mis_val_percent = 100 * mis_val / len(df)

    # Data type of columns with missing values
    mis_val_dtype = df.dtypes

    # Create a table with the results
    mis_val_table = pd.concat([mis_val, mis_val_percent, mis_val_dtype], axis=1)

    # Rename columns
    mis_val_table_ren_columns = mis_val_table.rename(
        columns={0: 'Missing Values', 1: '% of Missing Values', 2: 'Dtype'})

    # Sort the table by percentage of missing values in descending order and remove columns without missing values
    mis_val_table_ren_columns = mis_val_table_ren_columns[mis_val_table_ren_columns.iloc[:, 0] != 0].sort_values(
        '% of Missing Values', ascending=False).round(2)

    # Print
    print("The dataset has " + str(df.shape[1]) + " columns.\n"
          "Found: " + str(mis_val_table_ren_columns.shape[0]) + " columns with missing values.")

    if mis_val_table_ren_columns.shape[0] == 0:
        return

    # Return the dataframe with missing information
    return mis_val_table_ren_columns

In [69]:
# 27. Create table with missing values
df_missing = calculate_missing_values_by_column(dataset)

The dataset has 55 columns.
Found: 41 columns with missing values.


In [70]:
# 28. View
df_missing

Missing Values  % of Missing Values  \
Nb of sec with 37500B < Vol UL                    130254                86.84   
Nb of sec with 6250B < Vol UL < 37500B            111843                74.56   
Nb of sec with 125000B < Vol DL                    97538                65.02   
TCP UL Retrans. Vol (Bytes)                        96649                64.43   
Nb of sec with 31250B < Vol DL < 125000B           93586                62.39   
Nb of sec with 1250B < Vol UL < 6250B              92894                61.93   
Nb of sec with 6250B < Vol DL < 31250B             88317                58.88   
TCP DL Retrans. Vol (Bytes)                        88146                58.76   
HTTP UL (Bytes)                                    81810                54.54   
HTTP DL (Bytes)                                    81474                54.32   
Avg RTT DL (ms)                                    27829                18.55   
Avg RTT UL (ms)                                    27812                18.54   
Handset Manufacturer                                9559                 6.37   
Handset Type                                        9559                 6.37   
Last Location Name                                  1153                 0.77   
MSISDN/Number                                       1066                 0.71   
Bearer Id                                            991                 0.66   
Nb of sec with Vol UL < 1250B                        793                 0.53   
10 Kbps < UL TP < 50 Kbps (%)                        792                 0.53   
UL TP > 300 Kbps (%)                                 792                 0.53   
50 Kbps < UL TP < 300 Kbps (%)                       792                 0.53   
UL TP < 10 Kbps (%)                                  792                 0.53   
Nb of sec with Vol DL < 6250B                        755                 0.50   
250 Kbps < DL TP < 1 Mbps (%)                        754                 0.50   
50 Kbps < DL TP < 250 Kbps (%)                       754                 0.50   
DL TP < 50 Kbps (%)                                  754                 0.50   
DL TP > 1 Mbps (%)                                   754                 0.50   
IMEI                                                 572                 0.38   
IMSI                                                 570                 0.38   
Start                                                  1                 0.00   
End Offset (ms)                                        1                 0.00   
Total UL (Bytes)                                       1                 0.00   
Start Offset (ms)                                      1                 0.00   
End                                                    1                 0.00   
Avg Bearer TP DL (kbps)                                1                 0.00   
Dur (s)                                                1                 0.00   
Avg Bearer TP UL (kbps)                                1                 0.00   
Dur (ms)                                               1                 0.00   
Activity Duration UL (ms)                              1                 0.00   
Activity Duration DL (ms)                              1                 0.00   
Total DL (Bytes)                                       1                 0.00   

                                            Dtype  
Nb of sec with 37500B < Vol UL            float64  
Nb of sec with 6250B < Vol UL < 37500B    float64  
Nb of sec with 125000B < Vol DL           float64  
TCP UL Retrans. Vol (Bytes)               float64  
Nb of sec with 31250B < Vol DL < 125000B  float64  
Nb of sec with 1250B < Vol UL < 6250B     float64  
Nb of sec with 6250B < Vol DL < 31250B    float64  
TCP DL Retrans. Vol (Bytes)               float64  
HTTP UL (Bytes)                           float64  
HTTP DL (Bytes)                           float64  
Avg RTT DL (ms)                           float64  
Avg RTT UL (ms)

Typically, columns with more than 50% missing values should be removed. For those with 30% to 50% missing values, the decision is optional.

However, the final decision is always yours! Yes, you, Data Analyst. Just remember to always justify your choices.

In this project, we will remove columns with more than 30% missing values, as we have a large number of columns with missing values, which means a significant amount of work. We will handle variables with a low percentage of missing values and delete those with a high percentage of missing values.

### **1.2. Dropping Columns**

In [71]:
# 29. Columns to be removed
columns_to_remove = df_missing[df_missing['% of Missing Values'] >= 30.00].index.tolist()

In [72]:
# 30. Columns to be removed
columns_to_remove

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'TCP UL Retrans. Vol (Bytes)',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'TCP DL Retrans. Vol (Bytes)',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

Even though the “TCP” variables have many missing values, instead of removing them, we will apply imputation to these variables, as they may be necessary for our subsequent analysis.

In [73]:
# 31. Columns to be removed (excluding certain TCP variables)
columns_to_remove = [col for col in columns_to_remove if col not in ['TCP UL Retrans. Vol (Bytes)',
                                                                     'TCP DL Retrans. Vol (Bytes)']]

In [74]:
# 32. Columns to be removed
columns_to_remove

['Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'HTTP UL (Bytes)',
 'HTTP DL (Bytes)']

In [75]:
# 33. Drop columns and create a new dataframe
cleaned_dataset = dataset.drop(columns_to_remove, axis=1)

In [76]:
# 34. Shape
cleaned_dataset.shape

(150001, 47)

Now let’s check the status of the missing values in the modified dataframe.

In [77]:
# 35. Check the percentage of missing values in the modified dataframe
calculate_missing_values_percentage(cleaned_dataset)


The dataset has 3.85 % missing values.


In [78]:
# 36. Check missing values by column in the modified dataframe
calculate_missing_values_by_column(cleaned_dataset)


The dataset has 47 columns.
Found: 33 columns with missing values.


Missing Values  % of Missing Values    Dtype
TCP UL Retrans. Vol (Bytes)              96649                64.43  float64
TCP DL Retrans. Vol (Bytes)              88146                58.76  float64
Avg RTT DL (ms)                          27829                18.55  float64
Avg RTT UL (ms)                          27812                18.54  float64
Handset Type                              9559                 6.37   object
Handset Manufacturer                      9559                 6.37   object
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

### **1.3. Imputation with Backward Fill**

Since the percentages of missing values for 'TCP UL Retrans. Vol (Bytes)' and 'TCP DL Retrans. Vol (Bytes)' are very high, we will apply imputation to the missing values using the backward fill method.

In this case, using a single value like the mean or median is not advisable, as it may alter our data undesirably by making most values equal to a single value.

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html

In [79]:
# 37. Imputation of missing values using backward fill
# method = 'bfill': Bfill or backward-fill propagates the first observed non-null value backward until
# another non-null value is encountered.
def fix_missing_bfill(df, col):

    count = df[col].isna().sum()

    df[col] = df[col].fillna(method='bfill')

    print(f"{count} missing values in column {col} were replaced using the backward fill method.")


In [80]:
# 38. Backward Fill Imputation for the variable 'TCP UL Retrans. Vol (Bytes)'
fix_missing_bfill(cleaned_dataset, 'TCP UL Retrans. Vol (Bytes)')

96649 missing values in column TCP UL Retrans. Vol (Bytes) were replaced using the backward fill method.


In [81]:
# 39. Backward Fill Imputation for the variable 'TCP DL Retrans. Vol (Bytes)'
fix_missing_bfill(cleaned_dataset, 'TCP DL Retrans. Vol (Bytes)')

88146 missing values in column TCP DL Retrans. Vol (Bytes) were replaced using the backward fill method.


### **1.4. Imputation with Forward Fill**

In [82]:
# 40. Check missing values by column in the cleaned dataset
calculate_missing_values_by_column(cleaned_dataset)


The dataset has 47 columns.
Found: 33 columns with missing values.


Missing Values  % of Missing Values    Dtype
Avg RTT DL (ms)                          27829                18.55  float64
Avg RTT UL (ms)                          27812                18.54  float64
Handset Type                              9559                 6.37   object
Handset Manufacturer                      9559                 6.37   object
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
TCP UL Retrans. Vol (Bytes)                  5                 0.00  float64
TCP DL Retrans. Vol (Bytes)                  5                 0.00  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

"Avg RTT DL (ms)" and "Avg RTT UL (ms)" have the next highest percentages of missing values, with around 18.5% each. Let's check if these variables are skewed (do not follow a normal distribution) using the `skew()` method, which returns the skewness coefficient.

In [83]:
# 41. Check skewness for 'Avg RTT DL (ms)'
cleaned_dataset['Avg RTT DL (ms)'].skew(skipna=True)


62.90782807995961

In [84]:
# 42. Check skewness for 'Avg RTT UL (ms)'
cleaned_dataset['Avg RTT UL (ms)'].skew(skipna=True)

28.45741458546382

- Se a assimetria estiver entre -0,5 e 0,5, os dados são bastante simétricos
- Se a assimetria estiver entre -1 e - 0,5 ou entre 0,5 e 1, os dados estão moderadamente inclinados
- Se a assimetria for menor que -1 ou maior que 1, os dados estão altamente enviesados

Visto que ambas as colunas Avg RTT DL (ms) e Avg RTT UL (ms) são fortemente enviesadas positivamente é aconselhável não imputá-las com sua média. Portanto, usaremos o preenchimento progressivo.

In [85]:
# 43. Imputation of missing values using forward fill (progressive fill)
# method = 'ffill': Ffill or forward-fill propagates the last observed non-null value forward until
# another non-null value is encountered
def fix_missing_ffill(df, col):

    count = df[col].isna().sum()

    df[col] = df[col].fillna(method='ffill')

    print(f"{count} missing values in column {col} were replaced using the forward fill method.")


In [86]:
# 44. Forward Fill Imputation for 'Avg RTT DL (ms)'
fix_missing_ffill(cleaned_dataset, 'Avg RTT DL (ms)')


27829 missing values in column Avg RTT DL (ms) were replaced using the forward fill method.


In [87]:
# 45. Forward Fill Imputation for 'Avg RTT UL (ms)'
fix_missing_ffill(cleaned_dataset, 'Avg RTT UL (ms)')


27812 missing values in column Avg RTT UL (ms) were replaced using the forward fill method.


> We check the missing values again.

In [89]:
# 46. Check the percentage of missing values in the cleaned dataset
calculate_missing_values_percentage(cleaned_dataset)


The dataset has 0.44 % missing values.


In [90]:
# 47. Check missing values by column in the cleaned dataset
calculate_missing_values_by_column(cleaned_dataset)


The dataset has 47 columns.
Found: 31 columns with missing values.


Missing Values  % of Missing Values    Dtype
Handset Type                              9559                 6.37   object
Handset Manufacturer                      9559                 6.37   object
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
TCP UL Retrans. Vol (Bytes)                  5                 0.00  float64
TCP DL Retrans. Vol (Bytes)                  5                 0.00  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

### **1.5. Imputation of Categorical Variables**








In [91]:
# 48. Information about the cleaned dataset
cleaned_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 47 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Bearer Id                       149010 non-null  float64
 1   Start                           150000 non-null  object 
 2   Start Offset (ms)               150000 non-null  float64
 3   End                             150000 non-null  object 
 4   End Offset (ms)                 150000 non-null  float64
 5   Dur (s)                         150000 non-null  float64
 6   IMSI                            149431 non-null  float64
 7   MSISDN/Number                   148935 non-null  float64
 8   IMEI                            149429 non-null  float64
 9   Last Location Name              148848 non-null  object 
 10  Avg RTT DL (ms)                 150001 non-null  float64
 11  Avg RTT UL (ms)                 150001 non-null  float64
 12  Avg Bearer TP DL

Since "Handset Type" and "Handset Manufacturer" are categorical columns, it’s better to impute them with the value "unknown" to avoid biasing the data.

In [92]:
# 49. Fill NA values
def fix_missing_value(df, col, value):

    count = df[col].isna().sum()

    df[col] = df[col].fillna(value)

    if isinstance(value, str):
        print(f"{count} missing values in column {col} were replaced with '{value}'.")
    else:
        print(f"{count} missing values in column {col} were replaced with {value}.")


In [93]:
# 50. Imputation of categorical variables
fix_missing_value(cleaned_dataset, 'Handset Type', 'unknown')
fix_missing_value(cleaned_dataset, 'Handset Manufacturer', 'unknown')


9559 missing values in column Handset Type were replaced with 'unknown'.
9559 missing values in column Handset Manufacturer were replaced with 'unknown'.


We check the missing values again.

In [94]:
# 51. Check the percentage of missing values in the cleaned dataset
calculate_missing_values_percentage(cleaned_dataset)

The dataset has 0.17 % missing values.


In [95]:
# 52. Check missing values by column in the cleaned dataset
calculate_missing_values_by_column(cleaned_dataset)


The dataset has 47 columns.
Found: 29 columns with missing values.


Missing Values  % of Missing Values    Dtype
Last Location Name                        1153                 0.77   object
MSISDN/Number                             1066                 0.71  float64
Bearer Id                                  991                 0.66  float64
Nb of sec with Vol UL < 1250B              793                 0.53  float64
UL TP > 300 Kbps (%)                       792                 0.53  float64
50 Kbps < UL TP < 300 Kbps (%)             792                 0.53  float64
10 Kbps < UL TP < 50 Kbps (%)              792                 0.53  float64
UL TP < 10 Kbps (%)                        792                 0.53  float64
Nb of sec with Vol DL < 6250B              755                 0.50  float64
50 Kbps < DL TP < 250 Kbps (%)             754                 0.50  float64
DL TP > 1 Mbps (%)                         754                 0.50  float64
250 Kbps < DL TP < 1 Mbps (%)              754                 0.50  float64
DL TP < 50 Kbps (%)                        754                 0.50  float64
IMEI                                       572                 0.38  float64
IMSI                                       570                 0.38  float64
TCP UL Retrans. Vol (Bytes)                  5                 0.00  float64
TCP DL Retrans. Vol (Bytes)                  5                 0.00  float64
Start                                        1                 0.00   object
Avg Bearer TP UL (kbps)                      1                 0.00  float64
Avg Bearer TP DL (kbps)                      1                 0.00  float64
Dur (s)                                      1                 0.00  float64
End Offset (ms)                              1                 0.00  float64
Activity Duration DL (ms)                    1                 0.00  float64
Activity Duration UL (ms)                    1                 0.00  float64
Dur (ms)                                     1                 0.00  float64
End                                          1                 0.00   object
Start Offset (ms)                            1                 0.00  float64
Total UL (Bytes)                             1                 0.00  float64
Total DL (Bytes)                             1                 0.00  float64

### **1.6. Dropping Rows**

Since only 0.17% of the dataset contains missing values and the total number of rows is approximately 150,000, dropping these rows will not have a noticeable negative impact.

In [96]:
# 53. Drop rows with missing values
def drop_rows_with_na(df):

    old = df.shape[0]

    df.dropna(inplace=True)

    new = df.shape[0]

    count = old - new

    print(f"{count} rows containing missing values were dropped.")


In [97]:
# 54. Drop rows with missing values
drop_rows_with_na(cleaned_dataset)


3114 rows containing missing values were dropped.


In [98]:
# 55. Check the percentage of missing values in the cleaned dataset
calculate_missing_values_percentage(cleaned_dataset)

The dataset has 0.0 % missing values.


In [99]:
# 57. Shape
cleaned_dataset.shape

(146887, 47)

In [100]:
%watermark -a "panData"

Author: panData



In [101]:
%watermark

Last updated: 2024-11-01T07:40:29.367474+00:00

Python implementation: CPython
Python version       : 3.10.12
IPython version      : 7.34.0

Compiler    : GCC 11.4.0
OS          : Linux
Release     : 6.1.85+
Machine     : x86_64
Processor   : x86_64
CPU cores   : 2
Architecture: 64bit



In [102]:
%watermark --iversions

numpy     : 1.26.4
matplotlib: 3.8.0
sys       : 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
pandas    : 2.2.2



# **The End**